components/director/internal/open_resource_discovery/client.go

```diff
@@ -173,24 +170,47 @@ func buildDocumentURL(docURL, baseURL string) (string, error) { // TODO: tests w
 	return baseURL + docURL, nil
 }
 
-func buildWellKnownEndpoint(u string) (string, error) {
-	parsedURL, err := url.ParseRequestURI(u)
+func calculateBaseURL(webhookURL string, config WellKnownConfig) (string, error) {
+	parsedWebhookURL, err := url.ParseRequestURI(webhookURL)
 	if err != nil {
 		return "", errors.New("error while parsing input webhook url")
 	}
 
-	if parsedURL.Path != "" {
-		return parsedURL.String(), nil
+	hasWellKnownSuffix := strings.HasSuffix(parsedWebhookURL.Path, WellKnownEndpoint)
+
+	// if webhookURL is /well-known, strip the suffix and use it as baseURL. In case both are provided - the config baseURL is used.
+	// if webhookURL is not /well-known, but there is a valid baseURL provided in the config - use it
+	if hasWellKnownSuffix && config.BaseURL == "" {
+		strippedPath := strings.ReplaceAll(parsedWebhookURL.Path, WellKnownEndpoint, "")
+		parsedWebhookURL.Path = strippedPath
+		return parsedWebhookURL.String(), nil
+	} else if config.BaseURL != "" {
+		return config.BaseURL, nil
 	}
-	return parsedURL.String() + WellKnownEndpoint, nil
-}
 
-func stripRelativePathFromURL(u string) (string, error) {
-	parsedURL, err := url.ParseRequestURI(u)
+	// if webhookURL is not /well-known, there is no baseURL provided in the config - use the host of the absolute paths of the doc.URLs if it's the same for all of them
+	// we are sure that all docs have absolute URLs because if they didn't the Validate() of the config would have failed
+	documentHost, err := extractDocumentsURLHost(config.OpenResourceDiscoveryV1.Documents)
 	if err != nil {
-		return "", errors.New("error while parsing input webhook url")
+		return "", err
 	}
 
-	parsedURL.Path = ""
-	return parsedURL.String(), nil
+	return documentHost, nil
+}
+
+func extractDocumentsURLHost(docs []DocumentDetails) (string, error) {
```

Comment:
Regarding the second paragraph: if we reach `extractDocumentsURLHost` func, we are sure that all documents have absolute URLs in case no `/well-known` URL is passed or there is no `baseURL` provided in the config. If this is the case (no baseURL in both places) and there is at least one document with relative URL - we will throw validation error [here](https://github.com/kyma-incubator/compass/blob/ord-adjust-base-url/components/director/internal/open_resource_discovery/client.go#L48) when we validate the whole config object.

Which brings us to your first paragraph remark: I was wondering how to handle the "corner-case" when there is no `/well-known` URL passed or there is no `baseURL` provided in the config but every single document has an absolute URL. In that case, technically, we will be able to fetch the docs due to their absolute URLs but we still don't have a baseURL calculated that will be propagated in the further describedSystemInstance checks. So, after some discussions, a decision was made to extract the host of the docs.URL and check if it is the same for all of them - if not, validation error.

As anything besides the host can differ for each document, if there is a subpath on top of it, how can we be sure what part of that subpath is for the baseURL and what is for the specific doc - which brings us to the original issue that this task solves - we can not calculate the baseURL in that scenario. 

Do you have something in mind how to proceed in that particular scenario? Due to the fact that we can fetch the docs with their full URLs, I thought it would be incorrect from our side to throw a validation error, even though no baseURL was provied.

Reply:
i missed that check, for some reason - i guess i didn't expect that you would be processing/doing the well-known suffix check twice in two methods called one after the other. But you are right regarding the second point - i would still propose to simplify the code and remove some duplicate checks. Regarding the first point here is an idea - the spec says config baseURL is optional - this means you can keep it empty when there is no meaningful value to put in it (absolute docs + config not well-k + baseURL not returned by config response)

1.  First do some defaulting - calculate defaultBaseURL (based on presence of well-k suffix) only if c.BaseURL is empty & set c.BaseURL. This means call calculateBaseURL here but remove from it the last segment that sets base url based on absolute url host
2. Then call config.Validate, simplify the code there by removing the webhook host parsing and well-k checks from it
3. Then extend this check to make sure baseurl is not empty  if doc.DescribedSystemInstance != nil && doc.DescribedSystemInstance.BaseURL != nil && *doc.DescribedSystemInstance.BaseURL != baseURL
4. Then make sure to do the Sanitization with the baseurl coming from the doc itself